src/node.rs

```diff
@@ -0,0 +1,441 @@
+        hash: &EventHash,
+        round: usize,
+        unique_famous_witnesses: &BTreeSet<EventHash>
+    ) -> Result<(), Error> {
+        let timestamp_deciders = self.get_timestamp_deciders(hash, unique_famous_witnesses)?;
+        let times = timestamp_deciders.into_iter()
+            .map(|eh| self.hashgraph.get(&eh).unwrap().timestamp().unwrap())
+            .collect::<Vec<u64>>();
+        let times_sum: u64 = times.iter().sum();
+        let new_time = times_sum / times.len() as u64;
+        let event = self.hashgraph.get_mut(hash)?;
+        event.set_timestamp(new_time);
+        event.set_round_received(round);
+        Ok(())
+    }
+
+    #[inline]
+    fn get_timestamp_deciders(
+        &self,
+        hash: &EventHash,
+        unique_famous_witnesses: &BTreeSet<EventHash>
+    ) -> Result<BTreeSet<EventHash>, Error> {
+        let mut result = BTreeSet::new();
+        for unique_famous_witness in unique_famous_witnesses {
+            let self_ancestors = self.hashgraph.self_ancestors(unique_famous_witness).into_iter();
+            for self_ancestor in self_ancestors {
+                let ancestors = self.hashgraph.ancestors(self_ancestor);
+                let event = self.hashgraph.get(self_ancestor)?;
+                if ancestors.contains(&hash) && !event.is_self_parent(hash) {
+                    result.insert(self_ancestor.clone());
+                }
+            }
+        }
+        Ok(result)
+    }
+
+    #[inline]
+    fn get_unique_famous_witnesses(&self, round: usize) -> Result<BTreeSet<EventHash>, Error> {
+        let mut famous_witnesses = self.get_famous_witnesses(round);
+        for w in famous_witnesses.clone() {
+            for w1 in famous_witnesses.clone() {
+                if w != w1 {
+                    let e = self.hashgraph.get(&w)?;
+                    let e1 = self.hashgraph.get(&w1)?;
+                    if e.parents() == e1.parents() {
+                        famous_witnesses.remove(&w);
+                    }
+                }
+            }
+        }
+        Ok(famous_witnesses)
+    }
+
+    #[inline]
+    fn get_famous_witnesses(&self, round: usize) -> BTreeSet<EventHash> {
+        BTreeSet::from_iter(
+            self.rounds[round].witnesses().into_iter()
+                .filter(|eh| self.hashgraph.get(eh).unwrap().is_famous())
+        )
+    }
+
+    #[inline]
+    fn are_all_witnesses_famous(&self, round: usize) -> bool {
+        self.rounds[round].witnesses().iter()
+            .map(|eh| self.hashgraph.get(eh).unwrap())
+            .all(|e| e.is_famous())
+    }
+
+    #[inline]
+    fn get_vote(&self, witnesses: &BTreeSet<EventHash>, eh: &EventHash) -> (bool, usize) {
+        let mut total = 0;
+        for w in witnesses {
+            if self.votes[&(*w, *eh)] {
+                total += 1;
+            }
+        }
+        if total > witnesses.len()/2 {
+            (true, total)
+        } else {
+            (false, witnesses.len()-total)
```

Comment:
Do we need to `abs` or something here? - Just thinking of the condition where the length is less than total.

Reply:
It's not possible for it to be less than the total, because the total is sum from a subset of witnesses. But we could add abs to make that explicit.